[MB-3656] ADR: Deprecate use of lodash over time

suzubara opened the pull request:
## Description

This is an ADR describing a strategy for gradually deprecating use of lodash in our frontend code. It also adds the bundle analyzer recommended by Create React App for analyzing our static JS.

## References

* [Jira story](https://dp3.atlassian.net/browse/MB-3656) for this change

suzubara commented:
I'm looking for some initial feedback on the strategy I am proposing, and if accepted, I can amend this PR to add the optimization plugins and ESLint rules in a way that results in minimal initial code changes.

robot-mymove commented:
<!--
  0 failure: 
  0 warning: 
  1 messages
  
  DangerID: danger-id-Danger;
-->



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>:link: <a href="https://dp3.atlassian.net/browse/MB-3656">MB-3656</a></td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 8aaf81e05bc62f9e76eeeead60a965f27813571d
</p>

gidjin commented on docs/adr/0056-lodash.md:
```diff
@@ -0,0 +1,99 @@
+- Chosen Alternative: Deprecate lodash methods that have close native equivalents.
+- Additionally, set up the [lodash-webpack-plugin](https://www.npmjs.com/package/lodash-webpack-plugin) and [babel-plugin-lodash](https://www.npmjs.com/package/babel-plugin-lodash) to optimize how lodash adds to our bundle size.
+- `+` Enforce one consistent style throughout the codebase
+- `+` Can replace existing lodash methods on a per-method basis using the [ESLint plugin](https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore) (and prevent more instances from being added)
```
Plus one to automated enforcement! And bundle size reductions

gidjin commented on package.json:
```diff
@@ -83,6 +83,7 @@
   "name": "client",
   "private": true,
   "scripts": {
+    "analyze": "source-map-explorer 'build/static/js/*.js'",
```
This is neat!

suzubara commented on .spelling:
```diff
@@ -627,7 +627,32 @@ SwaggerClient
+natively
+transpiled
+polyfilled
 # Put all custom terms BEFORE this comment, lest 'pre-commit' and 'make spellcheck' yield different errors.
  - ./docs/backend/route-planner.md
  - docs/adr/0050-doc-viewer-fork.md
```
hm, I used the mdspell CLI and chose to ignore those words specifically in that file (instead of adding to the dictionary), so I assumed that was the syntax for that?

gidjin replied:
I think it is but I thought the list of ignored files needed to be at the end of the `.spelling` file

suzubara replied:
I think terms placed under each ignored file will be ignored specifically in those files (instead of ignoring the whole file)

gidjin replied:
Yeah I looked some more and remembered that `pre-commit` and `make spellcheck` can sometimes conflict if the code passes if terms are after the list of files to ignore, but if locally it's fine and build is good then maybe it's not an issue.

suzubara commented:
I modified this PR to include the following:
- Set up the ESLint plugin with an initial ruleset and fix resulting issues
- Added babel lodash plugin to cherry-pick lodash modules

I started setting up the webpack lodash plugin as well, but it requires you explicitly enable certain feature sets which I found obscure, so I think it would be best to wait to turn this on until we've cut down on usage.

lynzt commented:
I'm all for getting rid of dependencies, but I'm am wondering if we want to address this now or focus on other (higher priority) things...??

suzubara commented:
> I'm all for getting rid of dependencies, but I'm am wondering if we want to address this now or focus on other (higher priority) things...??

this came up in a previous F/E check-in, because our current inconsistent usage is leading to some confusion around which syntax is preferred. additionally, it is easier for us to identify potential areas of optimization earlier on rather than waiting (at which point we may have bought deeper into things like lodash and they will be that much harder to extract). that is why I am proposing this strategy, where we declare our intention to scale back on it now, but not enforce anything that requires significant code changes right away. I anticipate turning on more and more of the ESLint rules over time so that the whole team can acclimate to the native syntax gradually.

suzubara commented:
> I modified this PR to include the following:
> 
> * Set up the ESLint plugin with an initial ruleset and fix resulting issues
> * Added babel lodash plugin to cherry-pick lodash modules
> 
> I started setting up the webpack lodash plugin as well, but it requires you explicitly enable certain feature sets which I found obscure, so I think it would be best to wait to turn this on until we've cut down on usage.

One more change: I commented out the babel lodash plugin because something about using a custom babel config with rescripts is not working, and messing up our SVG imports. I will create follow up stories to figure this out but for now I would just like to get the ADR & lint rules in. @gidjin do you mind doing another pass since this now includes code changes?

gidjin commented on .rescriptsrc:
```diff
@@ -1,8 +1,7 @@
-    require.resolve('./case-sensitive-paths-rescripts.js'),
-  ],
-]
+  // ['use-babel-config', '.babelrc'], // TODO add babel lodash plugin
+  ['use-eslint-config', '.eslintrc'],
+  require.resolve('./terser-rescript.js'),
+  require.resolve('./case-sensitive-paths-rescripts.js'),
+  // require.resolve('./lodash-webpack-rescript.js') // TODO add this in once we've reduced lodash usage
```
This format is what we had incorrect before? Good find, glad we're actually calling our things now. How will we know we've reduced our lodash usage enough to enable the `lodash-webpack-rescripts` plugin?

suzubara replied:
the `lodash-webpack-plugin` works by completely [removing feature sets](https://github.com/lodash/lodash-webpack-plugin#feature-sets) from lodash (and things break if you remove something that you need), so I think once we have scoped down what we are using to a more understood subset, we can introduce this

gidjin commented on terser-rescript.js:
```diff
@@ -27,14 +27,6 @@ module.exports = (webpackConfig) => {
           },
           parallel: false,
           sourceMap: false,
-          chunkFilter: (chunk) => {
-            // Exclude uglification for the `vendor` chunk
-            if (chunk.name === 'vendor') {
-              return false;
-            }
-
-            return true;
-          },
```
Do you know why we needed to exclude uglification for vendor chunk? I'm not sure I grok that whole statement.

suzubara replied:
no I don't, but this was something I had added awhile back when looking into webpack build failures in CI that was taken from another github discussion somewhere, and fixing the rescripts syntax surfaced the fact that this option was deprecated (and had maybe never been implemented??) 🤷‍♀️

gidjin commented on docs/adr/0057-lodash.md:
```diff
@@ -0,0 +1,99 @@
+- [uniq](https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore#_uniq)
+- [without](https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore#_without)
+
+**Methods to retain (links to lodash docs):**
```
Is this and the above a comprehensive list? I wonder if we should explicitly call out if it is or not.

suzubara replied:
it's comprehensive to everything we're currently using from lodash (not the entirety of lodash). I imagine this list may evolve over time.. I wasn't sure if it made sense to include in the ADR itself but I wanted to have an easy place to point people to see what the alternative syntax was

gidjin replied:
I'm not opposed to updating this ADR's lists if we find more things to deprecate or retain.

gidjin commented on docs/adr/0057-lodash.md:
```diff
@@ -0,0 +1,99 @@
+## Decision Outcome
+
+- Chosen Alternative: Deprecate lodash methods that have close native equivalents.
+- Additionally, set up the [lodash-webpack-plugin](https://www.npmjs.com/package/lodash-webpack-plugin) and [babel-plugin-lodash](https://www.npmjs.com/package/babel-plugin-lodash) to optimize how lodash adds to our bundle size.
```
Are there existing stories or docs of the tech debt that note these still need to happen?

suzubara replied:
since it is currently unclear how to label this work in JIRA, I have added to a running tech debt doc I'm working on

gidjin reviewed (approved):
LGTM! I added a couple comments on areas I had questions but they are minor

jacquelineIO reviewed (approved):
looks good.
